### DISCUSSION TITLE
Deleting media in Download directory

### DISCUSSION THREAD

**Discussion by petazonk** [deckerst/aves#64] [Q&A] [answered]
I couldn't delete some videos from my download directory. Is this because the privileges have been minimised for security purposes?

**Comment by deckerst** [comment, marked as answer]
Yes, you are correct.

To strengthen user privacy and storage access control, Android 11 made the Download directory a [special directory with tighter access rules](https://developer.android.com/about/versions/11/privacy/storage#directory-access). For other directories, the user can give write access to the whole directory, but it's not possible with the Download directory.

I considered 2 solutions to this restriction, but they both have downsides:
- Solution 1: make Aves request an [additional permission](https://developer.android.com/about/versions/11/privacy/storage#all-files-access) which is reserved for file manager apps. This allows the app to access the Download directory just like other directories. But Aves is a gallery app, not a file manager app, so Google Play will not allow distributing Aves with this additional permission. I could only distribute Aves via other channels (like GitHub).
- Solution 2: request user consent for individual files in the Download directory. This is doable, but it would quickly become tedious for the user, and it would be inconsistent with the rest of the app.

**Reply by deckerst** [reply]
@petazonk since Aves v1.5.4 it is actually possible to modify, move or delete files in the Download directory on Android 11 and up. It will ask the user to give permission at the file level. It's a bit tedious, but there's no way around it.

**Reply by locuturus** [reply]
Is a follow up question cool? I appreciate this per-file permission work around even if it's clunky. I noticed it's not available in other directories, is this because you are interacting with the system downloads provider somehow?

**Reply by deckerst** [reply]
> Is a follow up question cool?

Sure, no worries.

> I appreciate this per-file permission work around even if it's clunky. I noticed it's not available in other directories,

In other directories, another kind of permission is used to let you edit the files. Are you saying that you can't edit the files at all in other directories?

> is this because you are interacting with the system downloads provider somehow?

No, it's not an issue of provider, but of permission. Media files are accessed differently depending on the Android version and file location:
- until `Android 10`, content on the device itself can be accessed by the `File API`, and content on removable storage by the `SAF API`,
- from `Android 11`, content on most of the storage can be accessed by the `SAF API` or the `Media Store API` (with some exceptions).

The `SAF API` is my preferred way of interacting with the files, because the app can request permission from the user for a whole subtree and the app doesn't have to bother the user again to edit any file in that subtree. However, this subtree permission request is not possible for some directories (most notably the `Downloads` folder, but also any volume root folder), so the app falls back to the `Media Store API`, which requires the per-file permission requests.

**Reply by locuturus** [reply]
Thanks for the reply!

> In other directories, another kind of permission is used to let you edit the files. Are you saying that you can't edit the files at all in other directories?

With SAF I can edit, yes. 

> The `SAF API` is my preferred way of interacting with the files, because the app can request permission from the user for a whole subtree and the app doesn't have to bother the user again to edit any file in that subtree. However, this subtree permission request is not possible for some directories (most notably the `Downloads` folder, but also any volume root folder), so the app falls back to the `Media Store API`, which requires the per-file permission requests.

So this per-file request is possible with `Media Store API` in other directories, but the `SAF API` experience is better so you only fall back to this method in `Downloads` or similarly restricted directories?

**Reply by deckerst** [reply]
Exactly.

There are also some edge cases where the `Media Store API` falls short. For example, creating a media file in a directory which is not within standard media subtrees (like `/DCIM/`, `/Pictures/`) is possible via `SAF API`, but not via `Media Store API`. For maximum compatibility, new albums created within Aves are placed within `/Pictures/`.
